app/src/main/cpp/HandMeshRenderer.cpp

```diff
@@ -0,0 +1,634 @@
+/* -*- Mode: C++; tab-width: 20; indent-tabs-mode: nil; c-basic-offset: 2 -*-
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "Controller.h"
+#include "DeviceUtils.h"
+#include "HandMeshRenderer.h"
+#include "tiny_gltf.h"
+
+#include "vrb/Camera.h"
+#include "vrb/Color.h"
+#include "vrb/ConcreteClass.h"
+#include "vrb/GLError.h"
+#include "vrb/ProgramFactory.h"
+#include "vrb/RenderState.h"
+#include "vrb/ShaderUtil.h"
+#include "vrb/Toggle.h"
+#include "vrb/gl.h"
+
+#define XR_EXT_HAND_TRACKING_NUM_JOINTS 26
+
+namespace crow {
+
+// HandMeshRendererSpheres
+
+struct HandMeshSpheres {
+    vrb::TogglePtr toggle;
+    std::vector<vrb::TransformPtr> sphereTransforms;
+};
+
+struct HandMeshRendererSpheres::State {
+    std::vector<HandMeshSpheres> handMeshState;
+};
+
+HandMeshRendererSpheres::HandMeshRendererSpheres(State &aState, vrb::CreationContextPtr& aContext)
+    : m(aState) {}
+
+HandMeshRendererSpheres::~HandMeshRendererSpheres() {}
+
+HandMeshRendererPtr HandMeshRendererSpheres::Create(vrb::CreationContextPtr& aContext) {
+    auto instance = (HandMeshRendererPtr) std::make_shared<vrb::ConcreteClass<HandMeshRendererSpheres, HandMeshRendererSpheres::State> >(aContext);
+    instance->Initialize(aContext);
+    return instance;
+}
+
+void HandMeshRendererSpheres::Update(Controller &aController, const vrb::GroupPtr& aRoot, const bool aEnabled) {
+    if (!aEnabled)
+        return;
```

Comment:
Actually, we can't remove the argument. There is a bug here, we need to be able to call `toggle->ToggleAll(false)` when `aEnabled`, otherwise we never hide the hands once they have been shown. I had this code like that and a later rewrite I changed it, but I remember there was a reason why I added the aEnable argument.

The key is that `Update()` needs to be called even when not on hand-tracking, because here is the only chance of hiding the mesh.